### ISSUE TITLE
BD-rate

### ISSUE THREAD

**Issue by lisongling33** [lyq133/LIC-HPCM#16]
Hello, thank you for your excellent work! Could you share the code you used to calculate BD-rate and rate-distortion curves? Additionally, could you provide the data for VTM-22.0 used in Kodak, Tecnik, and CLIC Pro Valid to calculate the six points (bpp, PSNR) corresponding to BD-rate.

**Comment by lyq133** [comment]
Hello, thank you for your interest. I have uploaded the following two files:

* `vtm22_rd_data.json`: the average VTM-22.0 bpp and RGB PSNR values for Kodak, Tecnick, and CLIC Pro Valid.
* `bd_rate.py`: the `BD_RATE` function used for our evaluation, adapted from [[Anserw/Bjontegaard_metric](https://github.com/Anserw/Bjontegaard_metric)](https://github.com/Anserw/Bjontegaard_metric).

To calculate the BD-rate of a tested method relative to VTM, you can use:

```python
BD_RATE(vtm_bpp, vtm_psnr, test_bpp, test_psnr)
```

Hope this helps!

[vtm22_rd_data.json](https://github.com/user-attachments/files/29889900/vtm22_rd_data.json)

[bd_rate.py](https://github.com/user-attachments/files/29889904/bd_rate.py)

**Comment by lisongling33** [comment]
Hello, thank you very much for sharing. I have a question about why there are only five sets of data for the bpp and PSNR of the Kodak dataset. Also, I would like to ask you, if I use two cards to run the experiment and adjust the code to the DDP version, will the configuration of this experiment change? For example, if I run 2 million steps with one card and a batch size of 32, and the training rounds are 3001, if I want to speed up the training and run it with two cards, will the batch size and training rounds change, or will they remain the same.

**Comment by lyq133** [comment]
> Hello, thank you very much for sharing. I have a question about why there are only five sets of data for the bpp and PSNR of the Kodak dataset. Also, I would like to ask you, if I use two cards to run the experiment and adjust the code to the DDP version, will the configuration of this experiment change? For example, if I run 2 million steps with one card and a batch size of 32, and the training rounds are 3001, if I want to speed up the training and run it with two cards, will the batch size and training rounds change, or will they remain the same.

Hello, thank you for your questions.

For the Kodak dataset, the number of RD points does not necessarily need to be the same for VTM and the tested model. BD-rate is calculated over the overlapping PSNR range of the two RD curves. In practice, the main consideration is whether the anchor and tested curves cover sufficiently aligned operating ranges, especially in terms of PSNR.

For DDP training, the important quantity is the global batch size:
`global batch size = batch size per GPU × number of GPUs`
If the original single-GPU experiment uses a batch size of 32, then for two-GPU DDP training, you should normally set the batch size to 16 per GPU, so that the global batch size remains 32. In this case, the learning rate, number of training epochs, and learning-rate schedule can remain unchanged.

**Comment by lisongling33** [comment]
Hello, thank you for your explanation. I have a set of BPP and PSNR data, and I want to measure BD-rate. Should I use the data with BPP1=[0.1097, 0.1720, 0.2719, 0.4098, 0.5874, 0.8215] and PSNR1=[29.1645, 30.6365, 32.3380, 34.1451, 35.9415, 37.7879] for these six points, or use the data with BPP2=[0.1097, 0.1720, 0.2719, 0.4098, 0.5874] and PSNR2=[29.1645, 30.6365, 32.3380, 34.1451, 35.9415] for these five points? The BD-rate obtained from these two sets of data differs by 0.4 points.

**Comment by lyq133** [comment]
> Hello, thank you for your explanation. I have a set of BPP and PSNR data, and I want to measure BD-rate. Should I use the data with BPP1=[0.1097, 0.1720, 0.2719, 0.4098, 0.5874, 0.8215] and PSNR1=[29.1645, 30.6365, 32.3380, 34.1451, 35.9415, 37.7879] for these six points, or use the data with BPP2=[0.1097, 0.1720, 0.2719, 0.4098, 0.5874] and PSNR2=[29.1645, 30.6365, 32.3380, 34.1451, 35.9415] for these five points? The BD-rate obtained from these two sets of data differs by 0.4 points.

I may suggest using the six-point data. The highest VTM point is approximately 0.801 bpp and 36.97 dB, while the five-point curve only reaches 0.587 bpp and 35.94 dB. The sixth point extends the tested curve to 0.822 bpp and 37.79 dB, which better covers the bitrate and PSNR range of the VTM anchor. The 0.4-point difference is caused by the change in the overlapping PSNR range.

**Comment by lisongling33** [comment]
Thank you for your reply. May I ask how long it takes you to train one point or complete the entire set of experiments for this experiment? I noticed that when you mentioned using a single 4090d graphics card and batch size of 8 for training, it takes approximately 2 to 3 days to complete 1.8 million steps and achieve competitive RD performance. Could you please tell me how this competitive RD performance compares to the original one? Also, after changing the batch size to 8, are there any learning rates that need to be adjusted? How should we adjust this part of the code? lr_scheduler = lambda x : \
 1e-4 if x < 2750 else (
 3e-5 if x < 2850 else (
 1e-5 if x < 2950 else 1e-6
 )
 )
It seems that there is a slight difference between it and the learning rate mentioned in the paper. Could you please advise me on whether it's better to use the learning rate from the paper or the one from the code? Also, I am very interested in your approach to experimental improvement. Would it be possible to discuss it with you later? Thank you very much for your selfless dedication and helpful answers.

**Comment by lyq133** [comment]
> Thank you for your reply. May I ask how long it takes you to train one point or complete the entire set of experiments for this experiment? I noticed that when you mentioned using a single 4090d graphics card and batch size of 8 for training, it takes approximately 2 to 3 days to complete 1.8 million steps and achieve competitive RD performance. Could you please tell me how this competitive RD performance compares to the original one? Also, after changing the batch size to 8, are there any learning rates that need to be adjusted? How should we adjust this part of the code? lr_scheduler = lambda x : 1e-4 if x < 2750 else ( 3e-5 if x < 2850 else ( 1e-5 if x < 2950 else 1e-6 ) ) It seems that there is a slight difference between it and the learning rate mentioned in the paper. Could you please advise me on whether it's better to use the learning rate from the paper or the one from the code? Also, I am very interested in your approach to experimental improvement. Would it be possible to discuss it with you later? Thank you very much for your selfless dedication and helpful answers.

Yes. With a single RTX 4090D and a batch size of 8, training one bitrate-point model usually takes about two to three days. Compared with the released models trained with a batch size of 32, the resulting performance may be approximately 1–2% worse in terms of BD-rate. This gap may be reduced by slightly extending the training, for example to around 2M+ steps if computational resources permit.

When changing the batch size to 8, the learning rate does not necessarily need to be adjusted, and you can continue using the original setting. The learning-rate schedule in the released code is similar to that described in the paper, and we observed comparable performance with both.

Feel free to contact me through GitHub issues or email if you would like to discuss this further.
